File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/file/exoplayer/specs/GivenAPlayingFile/WhenPlaybackCompletes.java
package com.lasthopesoftware.bluewater.client.playback.file.exoplayer.specs.GivenAPlayingFile;

import com.annimon.stream.Stream;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.Player;
import com.lasthopesoftware.bluewater.client.playback.file.PlayedFile;
import com.lasthopesoftware.bluewater.client.playback.file.PlayingFile;
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.ExoPlayerPlaybackHandler;

import org.junit.BeforeClass;
import org.junit.Test;
import org.mockito.stubbing.Answer;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;

public class WhenPlaybackCompletes {

	private static List<Player.EventListener> eventListeners = new ArrayList<>();

	private static PlayedFile playedFile;

	@BeforeClass
	public static void context() throws InterruptedException {
		final ExoPlayer mockExoPlayer = mock(ExoPlayer.class);
		doAnswer((Answer<Void>) invocation -> {
			eventListeners.add(invocation.getArgument(0));
			return null;
		}).when(mockExoPlayer).addListener(any());
		final ExoPlayerPlaybackHandler playbackHandler = new ExoPlayerPlaybackHandler(mockExoPlayer);

		final CountDownLatch countDownLatch = new CountDownLatch(1);
		playbackHandler
			.promisePlayback()
			.eventually(PlayingFile::promisePlayedFile)
			.then(p -> {
				playedFile = p;
				countDownLatch.countDown();
				return null;
			});

		Stream.of(eventListeners).forEach(e -> e.onPlayerStateChanged(false, Player.
			STATE_ENDED));

		countDownLatch.await();
	}

	@Test
	public void thenThePlayedFileIsReturned() {
		assertThat(playedFile).isNotNull();
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/file/exoplayer/GivenAPlayingFile/WhenPlaybackCompletes.kt
package com.lasthopesoftware.bluewater.client.playback.file.exoplayer.GivenAPlayingFile

import androidx.media3.common.Player
import com.lasthopesoftware.bluewater.client.playback.exoplayer.PromisingExoPlayer
import com.lasthopesoftware.bluewater.client.playback.file.exoplayer.ExoPlayerPlaybackHandler
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.lasthopesoftware.bluewater.shared.promises.extensions.toPromise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeUnit

class WhenPlaybackCompletes {

	private val eventListeners: MutableList<Player.Listener> = ArrayList()
	private val playedFile by lazy {
		val mockExoPlayer = mockk<PromisingExoPlayer>()
		every { mockExoPlayer.setPlayWhenReady(any()) } returns mockExoPlayer.toPromise()
		every { mockExoPlayer.removeListener(any()) } returns mockExoPlayer.toPromise()
		every { mockExoPlayer.addListener(any()) } answers {
			eventListeners.add(firstArg())
			mockExoPlayer.toPromise()
		}

		val playbackHandler = ExoPlayerPlaybackHandler(mockExoPlayer)

		val playedFileFuture =
			playbackHandler
				.promisePlayback()
				.eventually { obj -> obj.promisePlayedFile() }
				.toExpiringFuture()

		eventListeners.forEach { e -> e.onPlaybackStateChanged(Player.STATE_ENDED) }

		playedFileFuture[10, TimeUnit.SECONDS]
	}

	@Test
	fun `then the played file is returned`() {
		assertThat(playedFile).isNotNull
	}
}
